### PR TITLE
Add warning for legacy autograd function

### PR DESCRIPTION
**Pull request by yf225**
When working on https://github.com/pytorch/pytorch/pull/22762, we discovered that we haven't actually deprecated legacy autograd function. This PR puts up the deprecation warning for 1.2, with the goal to remove legacy function support completely in the near future.

### CODE REVIEW

File: torch/autograd/function.py [hunk 1]
```diff
@@ -98,6 +98,10 @@ def __init__(cls, name, bases, attrs):
 
         # old-style functions
         if not has_static_forward:
+            warnings.warn('Legacy autograd function with non-static forward method is deprecated. '
```
Comment (zou3519) [inline torch/autograd/function.py]: Add test?

File: torch/autograd/function.py [hunk 2]
```diff
@@ -98,6 +98,9 @@ def __init__(cls, name, bases, attrs):
...
```
Comment (zou3519) [inline torch/autograd/function.py]: I'm wondering if test_autograd.py will warning spew. At some point we should convert the legacy functions there that are not explicitly testing legacy behavior to non-legacy ones (probably doesn't have to happen in this PR)

Comment (yf225) [reply torch/autograd/function.py]: That's a good idea, I opened another PR for it: https://github.com/pytorch/pytorch/pull/22925.

**Comment by zou3519** [comment]
Test failure seems like it might be related?
```
Jul 16 21:05:30 ======================================================================
Jul 16 21:05:30 ERROR: test_legacy_function_deprecation_warning (__main__.TestAutograd)
Jul 16 21:05:30 ----------------------------------------------------------------------
Jul 16 21:05:30 Traceback (most recent call last):
Jul 16 21:05:30   File "test_autograd.py", line 180, in test_legacy_function_deprecation_warning
Jul 16 21:05:30     str(w[0].message))
Jul 16 21:05:30 IndexError: list index out of range
Jul 16 21:05:30 
```

**Comment by yf225** [comment]
@zou3519 I couldn't make the test work in Python 2.7, but I tested it manually and can confirm that it prints warning in Python 2.7.

**Comment by gchanan** [comment]
we have tests for warnings in python2 and this is a pretty major deprecation; we should have tests for it.

**Comment by yf225** [comment]
I moved the warning to the legacy function's forward method so that it can be reliably shown in both Python 2.7 and Python 3.

File: test/test_autograd.py [hunk 3]
```diff
@@ -167,6 +167,26 @@ def backward(self, grad_output):
         MyFunction()(y).sum().backward()
         self.assertEqual(v.grad.data, torch.zeros(shape))
 
+    def test_legacy_function_deprecation_warning(self):
+        with warnings.catch_warnings(record=True) as w:
+            # Ensure warnings are being shown
+            warnings.simplefilter("always")
```
Comment (zou3519) [inline test/test_autograd.py]: Is this necessary for this warning to be shown? I'm not sure what the default warnings behavior is

Comment (zou3519) [reply test/test_autograd.py]: Oh, I see why you had to do this:
```
print the first occurrence of matching warnings for each location (module + line number) where the warning is issued
```

File: torch/csrc/autograd/python_function.cpp [hunk 4]
```diff
@@ -631,6 +631,10 @@ PyObject *THPFunction_do_forward(THPFunction *self, PyObject *_inputs)
     std::vector<c10::IValue>(),
     autograd::Function::peek_at_next_sequence_nr());
 
+  TORCH_WARN("Legacy autograd function with non-static forward method is deprecated. ",
```
Comment (zou3519) [inline torch/csrc/autograd/python_function.cpp]: Maybe give a deadline if you want to remove it asap? "Will be removed in 1.3".

**Review by zou3519 — APPROVED** [review]
lgtm, I can't think of a use case where someone constructs an autograd function without calling it so putting a deprecation warning in forward should be fine.

**Pull request closed without merging**